src/Type/Data/Symbol.purs

```diff
@@ -45,3 +51,41 @@ instance equalsSymbol
 equals :: forall l r o. Equals l r o => SProxy l -> SProxy r -> BProxy o
 equals _ _ = BProxy
 
+-- | A type-level list of Symbols
+foreign import kind SList
+foreign import data SNil :: SList
+foreign import data SCons :: Symbol -> SList -> SList
+
+-- | Proxy specialised for SLists
+data SListProxy (l :: SList) = SListProxy
+
+-- | Sort a type-level list of Symbols in alphabetically ascending order
+class Sort (i :: SList) (o :: SList) | i -> o
+
+instance sortSNil  :: Sort SNil SNil
+instance sortSCons :: (Insert x ys xs', Sort xs ys) => Sort (SCons x xs) xs'
+
+-- Insert a Symbol into an ordered list
+class Insert (v :: Symbol) (i :: SList) (o :: SList) | i v -> o
+
+instance insertSNil :: Insert x SNil (SCons x SNil)
+instance insertSCons
+  :: ( CompareSymbol x y c
+     , InsertI c x y ys o
+     )
+  => Insert x (SCons y ys) o
+
+-- The actual insertion function, pattern matching on the result of the comparison
+class InsertI (c :: Ordering)
+              (x :: Symbol)
+              (y :: Symbol)
+              (ys :: SList)
+              (o :: SList) |
+              c x y ys -> o
+
+instance insertLT :: InsertI LT x y ys (SCons x (SCons y ys))
+instance insertGT :: InsertI GT x y ys (SCons y (SCons x ys))
```

Comment:
Don't you need to recurse and rebuild `ys` in the `GT` case?

Reply:
Ah, indeed!